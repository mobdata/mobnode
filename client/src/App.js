import React, { Component } from 'react' // basic building blocks of a React app
import PropTypes from 'prop-types' // typechecking on App input props
import { connect } from 'react-redux' // links App props to the store's state
import { renderRoutes } from 'react-router-config' // display various route endpoints (i.e. tabs)
import ClassificationBanner from '@mobdata/classification-banner' // displays classification of the host
import AppBar from 'components/AppBar' // MobNode banner
import TabMenu from 'components/TabMenu' // app navigation tabs
import { // utility actions
  getActiveTab, getAppVersion, changeRoute, changeTab,
} from 'modules/App'
import { toggleDrawer } from 'components/AppDrawer' // action for opening/closing drawer
import Button from '@material-ui/core/Button' // for toggling the drawer
import BuildIcon from '@material-ui/icons/Build' // wrench icon
import logErrorOnServer from 'util/serverLog' // show client errors server-side
import ErrorPage from 'components/ErrorPage' // displayed if the app crashes

/* Component which defines what to display for our application given the current route.

   Note that this is not the top-most component in our tree -- index.js is what actually
   defines application-level wrappers and setup (such as the Redux store). This component
   can be thought of as "Page" -- it is essentially a wrapper which defines the display on
   any given page or tab of the app.
*/
class App extends Component {
  // input prop typecheck
  static propTypes = {
    activeTab: PropTypes.string.isRequired,
    dispatch: PropTypes.func.isRequired,
    route: PropTypes.shape({
      routes: PropTypes.arrayOf(
        PropTypes.shape({}),
      ).isRequired,
    }).isRequired,
    history: PropTypes.shape({
      location: PropTypes.shape({
        pathname: PropTypes.string.isRequired,
      }).isRequired,
    }).isRequired,
  }

  constructor(props) {
    super(props)
    this.version = getAppVersion()
    this.state = {
      error: false, // set to true if a critical error bubbles up
    }
  }

  // on mount, return to the last page the user was on
  componentDidMount() {
    const { pathname } = this.props.history.location
    const lastIndex = pathname.lastIndexOf('/')
    let parsedPathname = pathname
    if (lastIndex > 0 && lastIndex < (pathname.length - 1)) {
      parsedPathname = pathname.substring(lastIndex)
    }
    this.props.dispatch(changeTab(parsedPathname.replace('/', '')))
  }

  // catch-all for errors not caught by a lower-level component
  componentDidCatch(error, info) {
    logErrorOnServer(
      1,
      'Captured by App-level componentDidCatch()',
      error,
      info.componentStack,
    )
    // triggers error page to display
    this.setState({ error: true })
  }

  // called on tab changes
  changeRoute = (path) => this.props.dispatch(changeRoute(path))

  // changes drawer type to match current tab on tab swap
  onMenuClick = () => {
    const { activeTab } = this.props
    let drawerType

    switch (activeTab) {
    case 'nodes':
      drawerType = 'NodeDrawer'
      break
      // add cases for drawers on other tabs here
    default:
      drawerType = ''
      break
    }

    this.props.dispatch(toggleDrawer(drawerType))
  }

  render() {
    const { routes } = this.props.route
    if (this.state.error) {
      return <ErrorPage error={this.state.error} />
    }
    return (
      <div>
        <ClassificationBanner />
        <div style={{ paddingTop: '25px' }}>
          <AppBar // Mobnode banner
            version={this.version}
          />
          <TabMenu // menu of clickable tabs to navigate the app
            activeTab={this.props.activeTab}
            onChange={this.changeRoute}
          />
          {/* Add names of any tabs which use a Drawer to this array */}
          { ['nodes'].includes(this.props.activeTab) && (
            <Button // wrench button which pulls out a drawer with utilities depending on the tab
              raised="true"
              dense="true"
              color="secondary"
              onClick={this.onMenuClick}
              style={{
                zIndex: 1000,
              }}
            >
              <BuildIcon />
            </Button>
          )}
          {renderRoutes(routes)}
        </div>
      </div>
    )
  }
}

const mapStateToProps = (state) => ({
  activeTab: getActiveTab(state),
})

export default connect(mapStateToProps)(App)
