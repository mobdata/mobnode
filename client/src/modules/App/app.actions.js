import { push } from 'connected-react-router'

export const CHANGE_TAB = 'CHANGE_TAB'
export const OPEN_DIALOG = 'OPEN_DIALOG'
export const CLOSE_DIALOG = 'CLOSE_DIALOG'

export const changeTab = (activeTab) => ({
  type: CHANGE_TAB,
  activeTab,
})

export const changeRoute = (path, changeActiveTab = true) => (dispatch) => {
  if (changeActiveTab) {
    dispatch(changeTab(path))
  }
  dispatch(push(path))
}


export const openDialog = (title, message, actions, list) => ({
  type: OPEN_DIALOG,
  actions,
  message,
  title,
  list,
  open: true,
})

export const closeDialog = () => ({
  type: CLOSE_DIALOG,
  open: false,
})
