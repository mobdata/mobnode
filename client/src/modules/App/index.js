export {
  CHANGE_TAB,
  changeTab,
  changeRoute,
  openDialog,
  closeDialog,
} from 'modules/App/app.actions'


export {
  default as AppReducer,
  getActiveTab,
  getAppVersion,
//  getDialogState,
} from 'modules/App/app.reducer'
