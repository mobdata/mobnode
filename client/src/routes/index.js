import React from 'react' // component rendering
import { Redirect } from 'react-router' // for redirecting to the rules page if the user goes to base /
import App from 'App' // specified to be rendered here

// the various page endpoints for our app
import EditDocPage from 'modules/EditDoc'
import NodePage from 'modules/Nodes'
import RulePage from 'modules/Rule'
import ConditionPage from 'modules/Conditions'
import HistoryPage from 'modules/History'
import Show404 from 'components/Show404'

// fetch any custom prefix to paths
const { REACT_APP_PATH_PREFIX } = process.env
const pathPrefix = (typeof REACT_APP_PATH_PREFIX !== 'undefined' && REACT_APP_PATH_PREFIX !== '')
  ? `/${REACT_APP_PATH_PREFIX}`
  : ''

// default to displaying the rule page if the user just navigates to the host itself
const defaultPath = `${pathPrefix}/rules`
const DefaultRoute = () => <Redirect to={defaultPath} />

const routes = [
  {
    // first renderRoutes(routes) will render our App and pass the routes array to App as a prop
    // App's renderRoutes(routes) then displays the page specified by the current path
    component: App,

    // defines what components to display at all URL endpoints of our app
    routes: [
      {
        path: `${pathPrefix}/`,
        exact: true,
        component: DefaultRoute,
      },
      {
        path: `${pathPrefix}/nodes`,
        exact: true,
        component: NodePage,
      },
      {
        path: `${pathPrefix}/rules`,
        exact: true,
        component: RulePage,
      },
      {
        path: `${pathPrefix}/conditions`,
        exact: true,
        component: ConditionPage,
      },
      {
        path: `${pathPrefix}/history`,
        exact: true,
        component: HistoryPage,
      },
      {
        path: `${pathPrefix}/editdoc`,
        exact: true,
        component: EditDocPage,
      },
      {
        // show a 404 page if the path doesn't match any of the above
        component: Show404,
      },
    ],
  },
]

export default routes
