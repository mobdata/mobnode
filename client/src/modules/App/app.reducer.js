import { CHANGE_TAB } from 'modules/App'

const initialState = {
  activeTab: 'rules',
}

const AppReducer = (state = initialState, action) => {
  const { type } = action

  switch (type) {
  case CHANGE_TAB:
    return { ...state, activeTab: action.activeTab || 'rules' }
  default:
    return state
  }
}

export const getActiveTab = (state) => state.app.activeTab

export const getAppVersion = () => process.env.REACT_APP_TAG

export default AppReducer
