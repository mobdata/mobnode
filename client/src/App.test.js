import React from 'react';
import ReactDOM from 'react-dom';
import { Provider, ReactReduxContext } from 'react-redux';
import Adapter from 'enzyme-adapter-react-16';// React 16 Enzyme adapter

const createHistory = require("history").createBrowserHistory
import { ConnectedRouter } from 'connected-react-router'
import Enzyme, { shallow, mount } from 'enzyme';
import toJson from 'enzyme-to-json';
import configureStore from './createStore'
const history = createHistory()
import routes from './routes' // defines which components render based on our current route/URL
import { renderRoutes } from 'react-router-config' // provides rendering-by-tab, i.e. routing


Enzyme.configure({ adapter: new Adapter() })
jest.mock('@mobdata/classification-banner', () => ({
  doSomething: jest.fn()
}))

const store = configureStore(history)
import App from './App';

describe('App mount tests', () => {

  beforeEach(() => {
    // Avoid `attachTo: document.body` Warning
    const div = document.createElement('div');
    div.setAttribute('id', 'container');
    document.body.appendChild(div);
  });

  afterEach(() => {
    const div = document.getElementById('container');
    if (div) {
      document.body.removeChild(div);
    }
  });

  test('renders defaultPath', () => {
    const pathPrefix = "";
    const defaultPath = `${pathPrefix}/rules`
    const DefaultRoute = () => <Redirect to={defaultPath} />
    const renderedComponent = {"compare":null,"displayName":"Connect(App)"}
    const renderedRoutes =  {"path":"/rules","exact":true, "component":{"propTypes":{},"displayName":"WithStyles(Connect(App))"}}
    const otherRoutes = {"routes": [
      {"path":"/nodes","exact":true,"component":{"propTypes":{},"displayName":"WithStyles(Connect(NodePage))"}},
      {"path":"/rules","exact":true, "component":{"propTypes":{},"displayName":"WithStyles(Connect(App))"}},
      {"path":"/conditions","exact":true,"component":{"propTypes":{},"displayName":"WithStyles(Connect(ConditionPage))"}},
      {"path":"/history","exact":true,"component":{"compare":null,"displayName":"Connect(HistoryPage)"}},
      {"path":"/editdoc","exact":true,"component":{"propTypes":{},"displayName":"WithStyles(EditDocPage)"}}
    ]}
    const renderedHistory = {"history":
                      {"length":1,"action":"REPLACE","location":
                       {"pathname":"/rules","search":"","hash":"","key":"h2zx6d"}
                      },
    "location":{"pathname":"/rules","search":"","hash":"","key":"h2zx6d"},
    "match":{"path":"/","url":"/","params":{},"isExact":false},
    "route":{"component":{"compare":null,"displayName":"Connect(App)"},
      "routes":[{"path":"/","exact":true},
        {"path":"/nodes","exact":true,"component":{"propTypes":{},"displayName":"WithStyles(Connect(NodePage))"}},
        {"path":"/rules","exact":true},
        {"path":"/conditions","exact":true,"component":{"propTypes":{},"displayName":"WithStyles(Connect(ConditionPage))"}},
        {"path":"/history","exact":true,"component":{"compare":null,"displayName":"Connect(HistoryPage)"}},
        {"path":"/editdoc","exact":true,"component":{"propTypes":{},"displayName":"WithStyles(EditDocPage)"}}
      ]
    },"activeTab":"rules"}

    const wrapper = shallow(
      <Provider store={store}>
        <ConnectedRouter history={history} >
          <App
            component= {renderedComponent}
            route= {renderedRoutes, otherRoutes}
            history= {renderedHistory}
          />
        </ConnectedRouter>
      </Provider>, { attachTo: document.getElementById('container') });

    expect(toJson(wrapper)).toMatchSnapshot();
  });

  test('renders nodePath', () => {
    const pathPrefix = "";
    const nodePath = `${pathPrefix}/nodes`
    const NodePage = () => <Redirect to={nodePath} />
    const wrapper = shallow(
      <Provider store={store}>
        <ConnectedRouter history={history} >
          <App
            route= {{ path: `/nodes`,
              exact: true,
              component: NodePage, }}
          />
        </ConnectedRouter>
      </Provider>, { attachTo: document.getElementById('container') });
    expect(toJson(wrapper)).toMatchSnapshot();

  });

  test('renders rulePath', () => {
    const pathPrefix = "";
    const rulePath = `${pathPrefix}/rules`
    const RulePage = () => <Redirect to={rulePath} />
    const wrapper = shallow(
      <Provider store={store}>
        <ConnectedRouter history={history} >
          <App
            route= {{ path: `/nodes`,
              exact: true,
              component: RulePage, }}
          />
        </ConnectedRouter>
      </Provider>, { attachTo: document.getElementById('container') });
    expect(toJson(wrapper)).toMatchSnapshot();

  });

  test('renders conditionPath', () => {
    const pathPrefix = "";
    const conditionPath = `${pathPrefix}/conditions`
    const ConditionPage = () => <Redirect to={conditionPath} />
    const wrapper = shallow(
      <Provider store={store}>
        <ConnectedRouter history={history} >
          <App
            route= {{ path: `/conditions`,
              exact: true,
              component: ConditionPage, }}
          />
        </ConnectedRouter>
      </Provider>, { attachTo: document.getElementById('container') });
    expect(toJson(wrapper)).toMatchSnapshot();

  });

  test('renders historyPath', () => {
    const pathPrefix = "";
    const historyPath = `${pathPrefix}/history`
    const HistoryPage = () => <Redirect to={historyPath} />
    const wrapper = shallow(
      <Provider store={store}>
        <ConnectedRouter history={history} >
          <App
            route= {{ path: `/history`,
              exact: true,
              component: HistoryPage, }}
          />
        </ConnectedRouter>
      </Provider>, { attachTo: document.getElementById('container') });
    expect(toJson(wrapper)).toMatchSnapshot();

  });

  test('renders Editdoc', () => {
    const pathPrefix = "";
    const editdocPath = `${pathPrefix}/editdoc`
    const EditDocPage = () => <Redirect to={editdocPath} />
    const wrapper = shallow(
      <Provider store={store}>
        <ConnectedRouter history={history} >
          <App
            route= {{ path: `/editdoc`,
              exact: true,
              component: EditDocPage, }}
          />
        </ConnectedRouter>
      </Provider>, { attachTo: document.getElementById('container') });
    expect(toJson(wrapper)).toMatchSnapshot();

  });

});
