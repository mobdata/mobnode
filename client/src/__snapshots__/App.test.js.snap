// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App mount tests renders Editdoc 1`] = `
<ContextProvider
  value={
    Object {
      "store": Object {
        "dispatch": [Function],
        "getState": [Function],
        "replaceReducer": [Function],
        "subscribe": [Function],
        Symbol(observable): [Function],
      },
      "subscription": Subscription {
        "handleChangeWrapper": [Function],
        "listeners": Object {
          "notify": [Function],
        },
        "onStateChange": [Function],
        "parentSub": undefined,
        "store": Object {
          "dispatch": [Function],
          "getState": [Function],
          "replaceReducer": [Function],
          "subscribe": [Function],
          Symbol(observable): [Function],
        },
        "unsubscribe": null,
      },
    }
  }
>
  <Connect(ConnectedRouterWithContext)
    history={
      Object {
        "action": "POP",
        "block": [Function],
        "createHref": [Function],
        "go": [Function],
        "goBack": [Function],
        "goForward": [Function],
        "length": 1,
        "listen": [Function],
        "location": Object {
          "hash": "",
          "pathname": "/",
          "search": "",
          "state": undefined,
        },
        "push": [Function],
        "replace": [Function],
      }
    }
  >
    <Connect(App)
      route={
        Object {
          "component": [Function],
          "exact": true,
          "path": "/editdoc",
        }
      }
    />
  </Connect(ConnectedRouterWithContext)>
</ContextProvider>
`;

exports[`App mount tests renders conditionPath 1`] = `
<ContextProvider
  value={
    Object {
      "store": Object {
        "dispatch": [Function],
        "getState": [Function],
        "replaceReducer": [Function],
        "subscribe": [Function],
        Symbol(observable): [Function],
      },
      "subscription": Subscription {
        "handleChangeWrapper": [Function],
        "listeners": Object {
          "notify": [Function],
        },
        "onStateChange": [Function],
        "parentSub": undefined,
        "store": Object {
          "dispatch": [Function],
          "getState": [Function],
          "replaceReducer": [Function],
          "subscribe": [Function],
          Symbol(observable): [Function],
        },
        "unsubscribe": null,
      },
    }
  }
>
  <Connect(ConnectedRouterWithContext)
    history={
      Object {
        "action": "POP",
        "block": [Function],
        "createHref": [Function],
        "go": [Function],
        "goBack": [Function],
        "goForward": [Function],
        "length": 1,
        "listen": [Function],
        "location": Object {
          "hash": "",
          "pathname": "/",
          "search": "",
          "state": undefined,
        },
        "push": [Function],
        "replace": [Function],
      }
    }
  >
    <Connect(App)
      route={
        Object {
          "component": [Function],
          "exact": true,
          "path": "/conditions",
        }
      }
    />
  </Connect(ConnectedRouterWithContext)>
</ContextProvider>
`;

exports[`App mount tests renders defaultPath 1`] = `
<ContextProvider
  value={
    Object {
      "store": Object {
        "dispatch": [Function],
        "getState": [Function],
        "replaceReducer": [Function],
        "subscribe": [Function],
        Symbol(observable): [Function],
      },
      "subscription": Subscription {
        "handleChangeWrapper": [Function],
        "listeners": Object {
          "notify": [Function],
        },
        "onStateChange": [Function],
        "parentSub": undefined,
        "store": Object {
          "dispatch": [Function],
          "getState": [Function],
          "replaceReducer": [Function],
          "subscribe": [Function],
          Symbol(observable): [Function],
        },
        "unsubscribe": null,
      },
    }
  }
>
  <Connect(ConnectedRouterWithContext)
    history={
      Object {
        "action": "POP",
        "block": [Function],
        "createHref": [Function],
        "go": [Function],
        "goBack": [Function],
        "goForward": [Function],
        "length": 1,
        "listen": [Function],
        "location": Object {
          "hash": "",
          "pathname": "/",
          "search": "",
          "state": undefined,
        },
        "push": [Function],
        "replace": [Function],
      }
    }
  >
    <Connect(App)
      component={
        Object {
          "compare": null,
          "displayName": "Connect(App)",
        }
      }
      history={
        Object {
          "activeTab": "rules",
          "history": Object {
            "action": "REPLACE",
            "length": 1,
            "location": Object {
              "hash": "",
              "key": "h2zx6d",
              "pathname": "/rules",
              "search": "",
            },
          },
          "location": Object {
            "hash": "",
            "key": "h2zx6d",
            "pathname": "/rules",
            "search": "",
          },
          "match": Object {
            "isExact": false,
            "params": Object {},
            "path": "/",
            "url": "/",
          },
          "route": Object {
            "component": Object {
              "compare": null,
              "displayName": "Connect(App)",
            },
            "routes": Array [
              Object {
                "exact": true,
                "path": "/",
              },
              Object {
                "component": Object {
                  "displayName": "WithStyles(Connect(NodePage))",
                  "propTypes": Object {},
                },
                "exact": true,
                "path": "/nodes",
              },
              Object {
                "exact": true,
                "path": "/rules",
              },
              Object {
                "component": Object {
                  "displayName": "WithStyles(Connect(ConditionPage))",
                  "propTypes": Object {},
                },
                "exact": true,
                "path": "/conditions",
              },
              Object {
                "component": Object {
                  "compare": null,
                  "displayName": "Connect(HistoryPage)",
                },
                "exact": true,
                "path": "/history",
              },
              Object {
                "component": Object {
                  "displayName": "WithStyles(EditDocPage)",
                  "propTypes": Object {},
                },
                "exact": true,
                "path": "/editdoc",
              },
            ],
          },
        }
      }
      route={
        Object {
          "routes": Array [
            Object {
              "component": Object {
                "displayName": "WithStyles(Connect(NodePage))",
                "propTypes": Object {},
              },
              "exact": true,
              "path": "/nodes",
            },
            Object {
              "component": Object {
                "displayName": "WithStyles(Connect(App))",
                "propTypes": Object {},
              },
              "exact": true,
              "path": "/rules",
            },
            Object {
              "component": Object {
                "displayName": "WithStyles(Connect(ConditionPage))",
                "propTypes": Object {},
              },
              "exact": true,
              "path": "/conditions",
            },
            Object {
              "component": Object {
                "compare": null,
                "displayName": "Connect(HistoryPage)",
              },
              "exact": true,
              "path": "/history",
            },
            Object {
              "component": Object {
                "displayName": "WithStyles(EditDocPage)",
                "propTypes": Object {},
              },
              "exact": true,
              "path": "/editdoc",
            },
          ],
        }
      }
    />
  </Connect(ConnectedRouterWithContext)>
</ContextProvider>
`;

exports[`App mount tests renders historyPath 1`] = `
<ContextProvider
  value={
    Object {
      "store": Object {
        "dispatch": [Function],
        "getState": [Function],
        "replaceReducer": [Function],
        "subscribe": [Function],
        Symbol(observable): [Function],
      },
      "subscription": Subscription {
        "handleChangeWrapper": [Function],
        "listeners": Object {
          "notify": [Function],
        },
        "onStateChange": [Function],
        "parentSub": undefined,
        "store": Object {
          "dispatch": [Function],
          "getState": [Function],
          "replaceReducer": [Function],
          "subscribe": [Function],
          Symbol(observable): [Function],
        },
        "unsubscribe": null,
      },
    }
  }
>
  <Connect(ConnectedRouterWithContext)
    history={
      Object {
        "action": "POP",
        "block": [Function],
        "createHref": [Function],
        "go": [Function],
        "goBack": [Function],
        "goForward": [Function],
        "length": 1,
        "listen": [Function],
        "location": Object {
          "hash": "",
          "pathname": "/",
          "search": "",
          "state": undefined,
        },
        "push": [Function],
        "replace": [Function],
      }
    }
  >
    <Connect(App)
      route={
        Object {
          "component": [Function],
          "exact": true,
          "path": "/history",
        }
      }
    />
  </Connect(ConnectedRouterWithContext)>
</ContextProvider>
`;

exports[`App mount tests renders nodePath 1`] = `
<ContextProvider
  value={
    Object {
      "store": Object {
        "dispatch": [Function],
        "getState": [Function],
        "replaceReducer": [Function],
        "subscribe": [Function],
        Symbol(observable): [Function],
      },
      "subscription": Subscription {
        "handleChangeWrapper": [Function],
        "listeners": Object {
          "notify": [Function],
        },
        "onStateChange": [Function],
        "parentSub": undefined,
        "store": Object {
          "dispatch": [Function],
          "getState": [Function],
          "replaceReducer": [Function],
          "subscribe": [Function],
          Symbol(observable): [Function],
        },
        "unsubscribe": null,
      },
    }
  }
>
  <Connect(ConnectedRouterWithContext)
    history={
      Object {
        "action": "POP",
        "block": [Function],
        "createHref": [Function],
        "go": [Function],
        "goBack": [Function],
        "goForward": [Function],
        "length": 1,
        "listen": [Function],
        "location": Object {
          "hash": "",
          "pathname": "/",
          "search": "",
          "state": undefined,
        },
        "push": [Function],
        "replace": [Function],
      }
    }
  >
    <Connect(App)
      route={
        Object {
          "component": [Function],
          "exact": true,
          "path": "/nodes",
        }
      }
    />
  </Connect(ConnectedRouterWithContext)>
</ContextProvider>
`;

exports[`App mount tests renders rulePath 1`] = `
<ContextProvider
  value={
    Object {
      "store": Object {
        "dispatch": [Function],
        "getState": [Function],
        "replaceReducer": [Function],
        "subscribe": [Function],
        Symbol(observable): [Function],
      },
      "subscription": Subscription {
        "handleChangeWrapper": [Function],
        "listeners": Object {
          "notify": [Function],
        },
        "onStateChange": [Function],
        "parentSub": undefined,
        "store": Object {
          "dispatch": [Function],
          "getState": [Function],
          "replaceReducer": [Function],
          "subscribe": [Function],
          Symbol(observable): [Function],
        },
        "unsubscribe": null,
      },
    }
  }
>
  <Connect(ConnectedRouterWithContext)
    history={
      Object {
        "action": "POP",
        "block": [Function],
        "createHref": [Function],
        "go": [Function],
        "goBack": [Function],
        "goForward": [Function],
        "length": 1,
        "listen": [Function],
        "location": Object {
          "hash": "",
          "pathname": "/",
          "search": "",
          "state": undefined,
        },
        "push": [Function],
        "replace": [Function],
      }
    }
  >
    <Connect(App)
      route={
        Object {
          "component": [Function],
          "exact": true,
          "path": "/nodes",
        }
      }
    />
  </Connect(ConnectedRouterWithContext)>
</ContextProvider>
`;
